File: backend/src/main/java/com/multi/finance/entity/Payment.java
package com.multi.finance.entity;

import com.multi.finance.enums.PaymentStatus;
import com.multi.finance.enums.PaymentType;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "payments")
public class Payment {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "bill_id", nullable = false)
    private Bill bill;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "group_id")
    private PaymentGroup group;

    @Column(nullable = false)
    private BigDecimal amount;

    @Enumerated(EnumType.STRING)
    @Column(name = "payment_type", nullable = false)
    private PaymentType paymentType;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PaymentStatus status;

    @Column(name = "reference_number")
    private String referenceNumber;

    @Column(name = "bank_name")
    private String bankName;

    @Column(name = "branch_name")
    private String branchName;

    @Column(name = "cheque_number")
    private String chequeNumber;

    @Column(name = "cheque_date")
    private LocalDate chequeDate;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "entered_by", nullable = false)
    private User enteredBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "confirmed_by")
    private User confirmedBy;

    @Column(name = "confirmed_at")
    private LocalDateTime confirmedAt;

    @Column(name = "payment_date", nullable = false)
    private LocalDate paymentDate;

    private String notes;

    @Column(name = "is_partial", nullable = false)
    private Boolean isPartial;

    @Column(name = "return_reason")
    private String returnReason;

    /**
     * The bill photographed when the payment was entered. Required of an accountant:
     * they are recording money already collected, and this is the only thing tying the
     * figure to the paper the customer signed.
     */
    @Column(name = "receipt_image_url", columnDefinition = "TEXT")
    private String receiptImageUrl;

    @Column(name = "receipt_uploaded_at")
    private LocalDateTime receiptUploadedAt;

    /**
     * The earlier payment this one's photograph was taken for.
     *
     * <p>Set when a customer settles a bill with two instruments in the same visit. Both
     * are written on the same page under one signature and one photograph shows both;
     * asking for a second would produce the same picture filed twice.
     *
     * <p>The URL itself is copied onto this payment, so every screen that shows a receipt
     * keeps working. This only records where the copy came from, which is what lets an
     * admin see that two payments rest on one photograph.
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "receipt_shared_from_payment_id")
    private Payment receiptSharedFrom;

    /**
     * The admin's own photograph, taken on confirmation. Optional and deliberately
     * separate: one records what the accountant saw, the other what the admin saw.
     */
    @Column(name = "confirm_image_url", columnDefinition = "TEXT")
    private String confirmImageUrl;

    @Column(name = "confirm_uploaded_at")
    private LocalDateTime confirmUploadedAt;

    // Set when this payment was pre-collected by the owner
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "collection_note_id")
    private CollectionNote collectionNote;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "collected_by_worker_id")
    private Worker collectedByWorker;

    @Column(name = "collector_note")
    private String collectorNote;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}

File: finance-frontend/src/app/core/services/payment.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface PaymentResponse {
  /** Photo taken by whoever entered the payment. */
  receiptImageUrl?: string;
  receiptUploadedAt?: string;
  /** The admin's own photo, attached on confirmation. */
  confirmImageUrl?: string;
  confirmUploadedAt?: string;

  id: number;
  billId: number;
  billNumber: string;
  billDate: string | null;
  customerName: string;
  business: string;
  billTotal: number;
  amountPaid: number;
  balanceRemaining: number;
  fullyPaid: boolean;
  paymentAmount: number;
  paymentType: string;
  status: string;
  isPartial: boolean;
  area: string | null;
  chequeNumber: string | null;
  chequeDate: string | null;
  bankName: string | null;
  branchName: string | null;
  referenceNumber: string | null;
  returnReason: string | null;
  enteredByName: string;
  confirmedByName: string | null;
  confirmedAt: string | null;
  paymentDate: string;
  notes: string | null;
  createdAt: string;
  collectionNoteId: number | null;
  collectedByOwnerName: string | null;
  collectedByOwnerAt: string | null;
  collectedByWorkerId: number | null;
  collectedByWorkerName: string | null;
  collectorNote: string | null;
}

export interface PaymentRequest {
  /** Photo of the bill. Required when an accountant enters the payment. */
  receiptImageUrl?: string;
  amount: number;
  paymentType: string;
  paymentDate?: string;
  chequeNumber?: string;
  chequeDate?: string;
  bankName?: string;
  branchName?: string;
  referenceNumber?: string;
  notes?: string;
}

export interface BulkPaymentBillItem {
  billId: number;
  amount: number;
  /**
   * Photo of this bill. Per bill, not per payment: one cheque may settle three bills,
   * but the three are separate pieces of paper signed separately.
   */
  receiptImageUrl?: string | null;
}

export interface BulkPaymentRequest {
  bills: BulkPaymentBillItem[];
  paymentType: string;
  chequeNumber?: string;
  bankName?: string;
  branchName?: string;
  chequeDate?: string;
  referenceNumber?: string;
  paymentDate?: string;
  notes?: string;
}

export interface PaymentGroupResponse {
  id: number;
  paymentType: string;
  chequeNumber?: string;
  bankName?: string;
  branchName?: string;
  referenceNumber?: string;
  chequeDate?: string;
  paymentDate: string;
  totalAmount: number;
  notes?: string;
  status: string;
  enteredByName: string;
  confirmedByName?: string;
  confirmedAt?: string;
  returnReason?: string;
  createdAt: string;
  payments: PaymentResponse[];
}

/**
 * Whether a bill still needs a photograph, and what already covers it if not.
 *
 * A photo the admin took when marking the collection, or one taken for an earlier
 * payment in the same handover, is evidence that already exists — asking again would
 * just produce the same picture twice.
 */
export interface ReceiptRequirement {
  /** Whether this user is asked for a photo at all. */
  required: boolean;
  /** Whether something already covers this bill. */
  canShare: boolean;
  /** How long a photo can cover a follow-up payment, in hours. */
  windowHours: number;
  sharedImageUrl?: string;
  /** Why it is covered, in words meant for the person entering. */
  sharedReason?: string;
  sharedFromPaymentId?: number;
  sharedFromAmount?: number;
  sharedFromAt?: string;

  /**
   * A collection the admin or owner marked and nobody has entered yet.
   *
   * The accountant is recording that same money, so the form starts from it rather than
   * being retyped from one panel into another — which is how a cash collection quietly
   * becomes a cheque and the two stop reconciling.
   */
  pendingNoteId?: number;
  pendingNoteAmount?: number;
  pendingNoteType?: 'CASH' | 'CHEQUE' | 'BANK_TRANSFER';
  pendingNoteBy?: string;
  pendingNoteAt?: string;
  pendingNoteChequeNumber?: string;
  pendingNoteBankName?: string;
  pendingNoteBranchName?: string;
  pendingNoteChequeDate?: string;
  pendingNoteReference?: string;
}

@Injectable({
  providedIn: 'root',
})
export class Payment {
  private apiUrl = `${environment.apiUrl}/payments`;

  constructor(private http: HttpClient) {}

  getAllPayments(status?: string, from?: string, to?: string): Observable<PaymentResponse[]> {
    let params = new HttpParams();
    if (status) params = params.set('status', status);
    if (from)   params = params.set('from', from);
    if (to)     params = params.set('to', to);
    return this.http.get<PaymentResponse[]>(this.apiUrl, { params });
  }

  enterPayment(billId: number, request: PaymentRequest): Observable<PaymentResponse> {
    return this.http.post<PaymentResponse>(`${this.apiUrl}/bills/${billId}`, request);
  }

  updatePayment(id: number, request: PaymentRequest): Observable<PaymentResponse> {
    return this.http.put<PaymentResponse>(`${this.apiUrl}/${id}`, request);
  }

  /** @param confirmImageUrl the admin's own photo, optional */
  confirmPayment(id: number, confirmImageUrl?: string): Observable<PaymentResponse> {
    return this.http.patch<PaymentResponse>(`${this.apiUrl}/${id}/confirm`,
      confirmImageUrl ? { confirmImageUrl } : {});
  }

  /**
   * Uploads a photo and returns its URL. Shares the task image endpoint — one place
   * that talks to ImageKit, rather than a second upload path to keep in step.
   */
  uploadImage(file: File): Observable<string> {
    const form = new FormData();
    form.append('file', file);
    return this.http.post<{ url: string }>(
      // Payments have their own ImageKit folder: a receipt is evidence for a figure
      // and is kept, where a task photo is not.
      `${environment.apiUrl}/payments/upload-image`, form
    ).pipe(map(r => r.url));
  }

  rejectPayment(id: number, reason: string): Observable<PaymentResponse> {
    return this.http.patch<PaymentResponse>(`${this.apiUrl}/${id}/reject`, { reason });
  }

  markChequeReturned(id: number, returnReason: string): Observable<PaymentResponse> {
    return this.http.patch<PaymentResponse>(`${this.apiUrl}/${id}/return`, { returnReason });
  }

  getPaymentsByBill(billId: number): Observable<PaymentResponse[]> {
    return this.http.get<PaymentResponse[]>(`${this.apiUrl}/bill/${billId}`);
  }

  getMyEnteredPayments(): Observable<PaymentResponse[]> {
    return this.http.get<PaymentResponse[]>(`${this.apiUrl}/my-entered`);
  }

  enterBulkPayment(request: BulkPaymentRequest): Observable<PaymentGroupResponse> {
    return this.http.post<PaymentGroupResponse>(`${this.apiUrl}/bulk`, request);
  }

  getGroups(status?: string): Observable<PaymentGroupResponse[]> {
    let params = new HttpParams();
    if (status) params = params.set('status', status);
    return this.http.get<PaymentGroupResponse[]>(`${this.apiUrl}/groups`, { params });
  }

  getGroup(id: number): Observable<PaymentGroupResponse> {
    return this.http.get<PaymentGroupResponse>(`${this.apiUrl}/groups/${id}`);
  }

  confirmGroup(id: number): Observable<PaymentGroupResponse> {
    return this.http.patch<PaymentGroupResponse>(`${this.apiUrl}/groups/${id}/confirm`, {});
  }

  returnGroup(id: number, returnReason: string): Observable<PaymentGroupResponse> {
    return this.http.patch<PaymentGroupResponse>(`${this.apiUrl}/groups/${id}/return`, { returnReason });
  }

  deletePayment(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getFutureCheques(customer?: string): Observable<PaymentResponse[]> {
    let params = new HttpParams();
    if (customer?.trim()) params = params.set('customer', customer.trim());
    return this.http.get<PaymentResponse[]>(`${this.apiUrl}/future-cheques`, { params });
  }

  searchByChequeNumber(chequeNumber: string): Observable<PaymentResponse[]> {
    return this.http.get<PaymentResponse[]>(`${this.apiUrl}/cheque-search`, {
      params: { chequeNumber },
    });
  }

  /** Read before the form is filled in, not after it is submitted. */
  getReceiptRequirement(billId: number): Observable<ReceiptRequirement> {
    return this.http.get<ReceiptRequirement>(
      `${this.apiUrl}/receipt-requirement/${billId}`);
  }
}